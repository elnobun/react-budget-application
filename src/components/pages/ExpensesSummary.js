import React from "react";
import { useSelector } from "react-redux";
import selectExpenseTotal from "../../redux/selectors/expenses-total";
import selectExpenses from "../../redux/selectors/expenses";
import numeral from "numeral";
import "numeral/locales/en-gb";

const ExpensesSummary = () => {
  // change locale currency
  numeral.locale("en-gb");

  // Connect component to store to get expenses and filters
  const visibleExpenses = useSelector((state) =>
    selectExpenses(state.expenses, state.filters)
  );

  // Get expenses length and the total expense amount
  const expenseCount = visibleExpenses.length;
  const expensesTotal = selectExpenseTotal(visibleExpenses);

  // Get expense word
  const expenseWord = expenseCount === 1 ? "expense" : "expenses";

  // Get expenseTotal and numeral value
  const total = numeral(expensesTotal / 100).format("$0,0.00");

  return (
    <div>
      <h2>
        Viewing {expenseCount} {expenseWord} totaling {total}
      </h2>
    </div>
  );
};

export default ExpensesSummary;
