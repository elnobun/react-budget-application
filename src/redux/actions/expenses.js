import { database } from "../../firebase/firebase";

// ADD EXPENSE
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense,
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const expense = {
      user_id: uid,
      description: expenseData.description || "",
      note: expenseData.note || "",
      amount: expenseData.amount || 0,
      createdAt: expenseData.createdAt || 0,
    };
    return database.add(expense).then((ref) => {
      dispatch(
        addExpense({
          id: ref.id,
          ...expense,
        })
      );
    });
  };
};

// REMOVE EXPENSE
export const removeExpense = (id = {}) => ({
  type: "REMOVE_EXPENSE",
  id,
});

export const startRemoveExpense = (id) => {
  return (dispatch) => {
    return database
      .doc(id)
      .delete()
      .then(() => {
        dispatch(removeExpense(id));
      });
  };
};

// EDIT EXPENSE
export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates,
});

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database
      .doc(id)
      .update(updates)
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};

// SET EXPENSES
export const setExpenses = (expenses) => ({
  type: "SET_EXPENSES",
  expenses,
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .where("user_id", "==", uid)
      .get()
      .then((querySnapshot) => {
        const expenses = [];

        querySnapshot.forEach((doc) => {
          expenses.push({
            id: doc.id,
            ...doc.data(),
          });
        });

        dispatch(setExpenses(expenses));
      });
  };
};
