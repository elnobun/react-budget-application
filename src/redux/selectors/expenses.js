// Get the visible expense
import moment from "moment";

const selectExpenses = (expenses, filters) => {
  // filter the expense array
  return expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      // Match start & end date
      const startDateMatch = filters.startDate
        ? filters.startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = filters.endDate
        ? filters.endDate.isSameOrAfter(createdAtMoment, "day")
        : true;

      // Match expenses description and filters text input
      const textMatch = expense.description
        .toLowerCase()
        .includes(filters.text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (filters.sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1; // descending order - lower 1st
      } else if (filters.sortBy === "amount") {
        return a.amount < b.amount ? 1 : -1; //descending order
      }
    });
};

export default selectExpenses;
