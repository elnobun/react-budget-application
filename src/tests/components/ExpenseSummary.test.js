import React from "react";
import ExpensesSummary from "../../components/pages/ExpensesSummary";
import { shallow } from "enzyme";
import { ReduxProvider } from "../setupProvider";
import expenses from "../fixtures/expenses";

test("should render ExpenseSummary with 1 expense", () => {
  const wrapper = shallow(
    <ReduxProvider>
      <ExpensesSummary expenseCount={1} expensesTotal={200} />
    </ReduxProvider>
  );
  expect(wrapper).toMatchSnapshot();
});

test("should render ExpenseSummary with multiple expense", () => {
  const wrapper = shallow(
    <ReduxProvider>
      <ExpensesSummary expenseCount={25} expensesTotal={25354678} />
    </ReduxProvider>
  );
  expect(wrapper).toMatchSnapshot();
});
