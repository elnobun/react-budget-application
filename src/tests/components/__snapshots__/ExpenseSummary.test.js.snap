// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseSummary with 1 expense 1`] = `
<Provider>
  <ExpensesSummary
    expenseCount={1}
    expensesTotal={200}
  />
</Provider>
`;

exports[`should render ExpenseSummary with multiple expense 1`] = `
<Provider>
  <ExpensesSummary
    expenseCount={25}
    expensesTotal={25354678}
  />
</Provider>
`;
